crates/nu-engine/src/eval_ir.rs

```diff
@@ -352,7 +352,10 @@ fn eval_instruction<D: DebugContext>(
         }
         Instruction::StoreVariable { var_id, src } => {
             let value = ctx.collect_reg(*src, *span)?;
-            ctx.stack.add_var(*var_id, value);
+            // Perform runtime type checking and conversion for variable assignment
+            let variable = ctx.engine_state.get_var(*var_id);
+            let converted_value = check_assignment_type(value, &variable.ty)?;
```

Comment:
Should `value` be taken by reference now? I wonder if we would do any modification here after these discussions: https://github.com/nushell/nushell/pull/16079#issuecomment-3040304253

Reply:
I'll leave it as is then

Reply:
I'd go for the second signature above (`Result<(), ShellError>`), since we're not actually doing any conversions right now. It might confuse someone to see the `/// Type check and convert value` doc comment and the return type. Not a huge deal though